In [1]:
import utils
import os
import pandas as pd
import sklearn
from sklearn import *
import os
import errno
import pickle
import numpy as np

# Simple model (Count Vectorizer)

In [2]:
path_folder_quora = "./Datasets/QuoraQuestionPairs/"

In [3]:
train_df = pd.read_csv(os.path.join(path_folder_quora, "quora_train_data.csv"))
A_df, te_df = sklearn.model_selection.train_test_split(train_df,test_size=0.05,random_state=123)
tr_df, va_df = sklearn.model_selection.train_test_split(A_df,test_size=0.05,random_state=123)
print('tr_df.shape=',tr_df.shape)
print('va_df.shape=',va_df.shape)
print('te_df.shape=',te_df.shape)

tr_df.shape= (291897, 6)
va_df.shape= (15363, 6)
te_df.shape= (16172, 6)


In [4]:
tr_df.head()

id    qid1    qid2  \
61482   125898  203030  203031   
131546   36249   66113   66114   
22927   199864  301469  301470   
183520  277339   17728  138400   
67694   392907  525647  525648   

                                                question1  \
61482   Is Java or C++ or C the most popular language ...   
131546  How do you convert direct speech into reported...   
22927                  Where can I buy used wine barrels?   
183520  What was the best day of your life? (Excluding...   
67694                           How is web-work.in works?   

                                                question2  is_duplicate  
61482   How do I develop a software which will have a ...             0  
131546  I feel weak at spoken English. I have sentence...             0  
22927                Where can you buy used wine barrels?             1  
183520       What is the Best Day of your life till date?             1  
67694                    How do I get web designing work?             0

In [5]:
q1_train =  utils.cast_list_as_strings(list(tr_df["question1"]))
q2_train =  utils.cast_list_as_strings(list(tr_df["question2"]))
q1_validation =  utils.cast_list_as_strings(list(va_df["question1"]))
q2_validation =  utils.cast_list_as_strings(list(va_df["question2"]))

In [6]:
q1_train[0], q2_train[0]

('Is Java or C++ or C the most popular language amongst startups for backend development?',
 'How do I develop a software which will have a Java GUI and a C++ or C backend?')

In [7]:
all_train_questions = q1_train + q2_train

count_vectorizer = sklearn.feature_extraction.text.CountVectorizer(ngram_range=(1,1))
count_vectorizer.fit(all_train_questions)

In [ ]:
X_tr_q1q2 = utils.get_features_from_df(tr_df, count_vectorizer)

X_tr_q1q2.shape, tr_df.shape

((291897, 149650), (291897, 6))

In [ ]:
logistic = sklearn.linear_model.LogisticRegression(solver="liblinear", random_state=123)
y_train = tr_df["is_duplicate"].values
logistic.fit(X_tr_q1q2, y_train)

LogisticRegression(random_state=123, solver='liblinear')

In [10]:
mistake_indices_tr, predictions_tr = utils.get_mistakes(logistic, X_tr_q1q2,  y_train)

utils.print_mistake_k(10, tr_df, mistake_indices_tr, predictions_tr)

What is a good website for free ebooks?
What are the sites to download free eBooks?
true class: 1
prediction: 0


In [11]:
# check if directory exists
if not os.path.exists('model_artifacts'):
    try:
        # create folder
        os.makedirs('model_artifacts')
        # open file descriptors for writing
        countvectorizer_file = open('model_artifacts/count_vectorizer.pkl', 'wb')
        logistic_file = open('model_artifacts/logistic.pkl', 'wb')
        # serialize and write objects to files
        pickle.dump(count_vectorizer, countvectorizer_file)
        pickle.dump(logistic, logistic_file)
        # close file descriptors
        countvectorizer_file.close()
        logistic_file.close()
    except OSError as e:
        if e.errno != errno.EEXIST:
            raise

In [12]:
# w1vec = word2vec.transform(w1)
# len(w1vec) = 256

# w2vec = word2vec.transform(w2)
# len(w2vec) = 256

# EmbA  = w1vec + ... + wnvec -> 256
# EmbB  = v1vec + ... + vnvec -> 256
# concat(256, 256)            -> 512
# word2vec feature: d(EmbA, EmbB)
# X_q1q2
# X_q1q2_fMaria_fPedro_fAdrian -> XGBoost / Logistic

# Improved model

In [2]:
path_folder_quora = "./Datasets/QuoraQuestionPairs/"

In [3]:
train_df = pd.read_csv(os.path.join(path_folder_quora, "quora_train_data.csv"))
A_df, te_df = sklearn.model_selection.train_test_split(train_df,test_size=0.05,random_state=123)
tr_df, va_df = sklearn.model_selection.train_test_split(A_df,test_size=0.05,random_state=123)
print('tr_df.shape=',tr_df.shape)
print('va_df.shape=',va_df.shape)
print('te_df.shape=',te_df.shape)

tr_df.shape= (291897, 6)
va_df.shape= (15363, 6)
te_df.shape= (16172, 6)


In [4]:
tr_df.head()

id    qid1    qid2  \
61482   125898  203030  203031   
131546   36249   66113   66114   
22927   199864  301469  301470   
183520  277339   17728  138400   
67694   392907  525647  525648   

                                                question1  \
61482   Is Java or C++ or C the most popular language ...   
131546  How do you convert direct speech into reported...   
22927                  Where can I buy used wine barrels?   
183520  What was the best day of your life? (Excluding...   
67694                           How is web-work.in works?   

                                                question2  is_duplicate  
61482   How do I develop a software which will have a ...             0  
131546  I feel weak at spoken English. I have sentence...             0  
22927                Where can you buy used wine barrels?             1  
183520       What is the Best Day of your life till date?             1  
67694                    How do I get web designing work?             0

## Preprocessing

First, a preprocessing is applied to the 2 questions: Casting it as strings, lowering and removing stop words. We also tokenize them.

In [5]:
#Cast
q1_train_c =  utils.cast_list_as_strings(list(tr_df["question1"]))
q2_train_c =  utils.cast_list_as_strings(list(tr_df["question2"]))
q1_validation_c =  utils.cast_list_as_strings(list(va_df["question1"]))
q2_validation_c =  utils.cast_list_as_strings(list(va_df["question2"]))

In [6]:
#Lower
q1_train_lw = utils.lower_list(q1_train_c)
q2_train_lw = utils.lower_list(q2_train_c)
q1_validation_lw = utils.lower_list(q1_validation_c)
q2_validation_lw = utils.lower_list(q2_validation_c)

In [7]:
#Remove stop words
stop_words = ["a", "an", "the", "and", "but", "or", "in", "on", "at", "to", "of", "for","i","you","he","she","it","we","they",
             "me","him","her","us","them","my","your","his","its","our","their","mine","yours","hers","ours","theirs","myself",
              "yourself","himself","herself","itself","ourselves","yourselves","themselves","this","that","these","those"]
q1_train = utils.remove_sw(q1_train_lw,stop_words)
q2_train = utils.remove_sw(q2_train_lw,stop_words)
q1_validation = utils.remove_sw(q1_validation_lw,stop_words)
q2_validation = utils.remove_sw(q2_validation_lw,stop_words)

In [8]:
#Tokenize
q1_train_tokens = utils.tokenize(q1_train)
q2_train_tokens = utils.tokenize(q2_train)
q1_validation_tokens = utils.tokenize(q1_validation)
q2_validation_tokens = utils.tokenize(q2_validation)

In [9]:
print(len(q1_train_tokens),len(q2_train_tokens))

291897 291897


## Jaccard distance at sentence level

In [10]:
jd_feature_train = utils.generate_jd_feature(q1_train_tokens,q2_train_tokens)
jd_feature_val = utils.generate_jd_feature(q1_validation_tokens,q2_validation_tokens)

In [11]:
jd_feature_train = np.array(jd_feature_train).reshape(-1,1)
jd_feature_train.shape

(291897, 1)

In [12]:
print(q1_train_tokens[2])
print(q2_train_tokens[2])
print(tr_df["is_duplicate"].values[2])
print(jd_feature_train[2])

['where', 'can', 'buy', 'used', 'wine', 'barrels']
['where', 'can', 'buy', 'used', 'wine', 'barrels']
1
[0.]


In [13]:
print(q1_train_tokens[3])
print(q2_train_tokens[3])
print(tr_df["is_duplicate"].values[3])
print(jd_feature_train[3])

['what', 'was', 'best', 'day', 'life', 'excluding', 'family', 'things', 'like', 'births']
['what', 'is', 'best', 'day', 'life', 'till', 'date']
1
[0.69230769]


## Unit tests TF-IDF and Cosine Distatnce

In [14]:
import tests

tests.test_cosine_distance()
tests.test_tfidf_vectorizer()
tests.test_build_w2v_model()
tests.test_w2v_embedding()

AttributeError: 'TfidfVectorizer' object has no attribute 'toarray'

In [15]:
vectorizer = utils.tfidf_vectorizer(all_train_questions, max_features=None)

tfidf_vectorizer_train_q1 = vectorizer.transform(q1_train)
tfidf_vectorizer_train_q2 = vectorizer.transform(q2_train)

tfidf_vectorizer_val_q1 = vectorizer.transform(q1_validation)
tfidf_vectorizer_val_q2 = vectorizer.transform(q2_validation)

In [31]:
import scipy
concat_train = scipy.sparse.hstack((tfidf_vectorizer_train_q1,tfidf_vectorizer_train_q2))
concat_val = scipy.sparse.hstack((tfidf_vectorizer_val_q1,tfidf_vectorizer_val_q2))

In [ ]:
cosine_distance_train = np.array([utils.cosine_distance(i,j) for i, j in zip(tfidf_vectorizer_train_q1,tfidf_vectorizer_train_q2)]).reshape(-1,1)
cosine_distance_val = np.array([utils.cosine_distance(i,j) for i, j in zip(tfidf_vectorizer_val_q1,tfidf_vectorizer_val_q2)]).reshape(-1,1)

cosine_distance_train = np.nan_to_num(cosine_distance_train, nan=0.0)
cosine_distance_val = np.nan_to_num(cosine_distance_val, nan=0.0)

/Users/eyuelmelese/Desktop/master/NLP/NLP-Quora/utils.py:112: RuntimeWarning: invalid value encountered in divide
  vector1.T/np.linalg.norm(vector1),
/Users/eyuelmelese/Desktop/master/NLP/NLP-Quora/utils.py:113: RuntimeWarning: invalid value encountered in divide
  vector2.T/np.linalg.norm(vector2))


## Comparing Words (first one and ordered)

In [17]:
fw_feature_train = utils.generate_first_word_feature(q1_train_tokens,q2_train_tokens)
fw_feature_val = utils.generate_first_word_feature(q1_validation_tokens,q2_validation_tokens)

In [18]:
fw_feature_train = np.array(fw_feature_train).reshape(-1,1)
fw_feature_train.shape

(291897, 1)

In [19]:
print(q1_train_tokens[2])
print(q2_train_tokens[2])
print(tr_df["is_duplicate"].values[2])
print(fw_feature_train[2])

['where', 'can', 'buy', 'used', 'wine', 'barrels']
['where', 'can', 'buy', 'used', 'wine', 'barrels']
1
[1]


In [20]:
print(q1_train_tokens[3])
print(q2_train_tokens[3])
print(tr_df["is_duplicate"].values[3])
print(fw_feature_train[3])

['what', 'was', 'best', 'day', 'life', 'excluding', 'family', 'things', 'like', 'births']
['what', 'is', 'best', 'day', 'life', 'till', 'date']
1
[1]


In [21]:
ow_feature_train = utils.generate_ordered_words_feature(q1_train_tokens,q2_train_tokens)
ow_feature_val = utils.generate_ordered_words_feature(q1_validation_tokens,q2_validation_tokens)

ow_feature_train = np.array(ow_feature_train).reshape(-1,1)
ow_feature_train.shape

(291897, 1)

In [22]:
print(q1_train_tokens[2])
print(q2_train_tokens[2])
print(tr_df["is_duplicate"].values[2])
print(ow_feature_train[2])
print(q1_train_tokens[3])
print(q2_train_tokens[3])
print(tr_df["is_duplicate"].values[3])
print(ow_feature_train[3])

['where', 'can', 'buy', 'used', 'wine', 'barrels']
['where', 'can', 'buy', 'used', 'wine', 'barrels']
1
[1.]
['what', 'was', 'best', 'day', 'life', 'excluding', 'family', 'things', 'like', 'births']
['what', 'is', 'best', 'day', 'life', 'till', 'date']
1
[0.57142857]


## Negation

In [23]:
neg_feature_train = utils.generate_negation_feature(q1_train,q2_train)
neg_feature_val = utils.generate_negation_feature(q1_validation,q2_validation)

neg_feature_train = np.array(neg_feature_train).reshape(-1,1)
neg_feature_train.shape

100%|██████████| 15363/15363 [03:10<00:00, 80.77it/s] 


(291897, 1)

In [32]:
print(q1_train[2])
print(q2_train[2])
print(tr_df["is_duplicate"].values[2])
print(neg_feature_train[2])

where can  buy used wine barrels?
where can  buy used wine barrels?
1
[1]


In [33]:
print(q1_train[27])
print(q2_train[27])
print(tr_df["is_duplicate"].values[27])
print(neg_feature_train[27])

can someone still get  message  snapchat if  haven't read  yet   deactivated  account?
if  block someone  snapchat will  still see  last message  sent?
0
[0]


## Training a logistic model

In [40]:
X_tr.shape

(291897, 149651)

In [37]:
#Concatenate here all your features, bros
X_tr = scipy.sparse.hstack((jd_feature_train, concat_train))

In [38]:
logistic_improved = sklearn.linear_model.LogisticRegression(solver="liblinear", random_state=123)
y_train = tr_df["is_duplicate"].values
logistic_improved.fit(X_tr, y_train)

LogisticRegression(random_state=123, solver='liblinear')

In [78]:
metrics.accuracy_score(y_train, logistic_improved.predict(X_tr))

0.8201591657331182

In [45]:
jd_feature_val = np.array(jd_feature_val).reshape(-1,1)
fw_feature_val = np.array(fw_feature_val).reshape(-1,1)
ow_feature_val = np.array(ow_feature_val).reshape(-1,1)
neg_feature_val = np.array(neg_feature_val).reshape(-1,1)

In [79]:
X_val = np.concatenate((jd_feature_val, cosine_distance_val, fw_feature_val, ow_feature_val), axis=1)
y_val = va_df["is_duplicate"].values

In [80]:
metrics.accuracy_score(y_val, logistic_improved.predict(X_val))

0.6457072186421923

Validation results

Sin fw_feature daba 0.6403697194558354 (pequeña ganancia) \
Con fw_feature da 0.6404999023628197 \
Si añadimos ow_feature sube a 0.6457072186421923 \
Si añadimos neg_feature sube a 0.6472694135260041 (aunque se necesita 1 hora para obtener el feature por lo que se descarta)

In [81]:
try:
    # open file descriptor for writing
    logistic_file = open('model_artifacts/logistic_imp.pkl', 'wb')
    # serialize and write object to file
    pickle.dump(logistic_improved, logistic_file)
    # close file descriptor
    logistic_file.close()
except OSError as e:
    if e.errno != errno.EEXIST:
        raise